### PR TITLE
[cross_file] An abstraction to allow working with files across multiple platforms.

### PR DESCRIPTION
**Pull request by mvanbeusekom**
## Description

This PR copies the [x_file](https://github.com/flutter/plugins/tree/master/packages/file_selector/file_selector_platform_interface/lib/src/types/x_file) type from the [file_selector](https://github.com/flutter/plugins/tree/master/packages/file_selector) package into its own package. This will allow us to easily reuse this cross platform file abstraction in other plugins that are dealing with files, without having to duplicate the code.

At the moment there is a requirement to use this type in the [camera](https://github.com/flutter/plugins/tree/master/packages/camera) plugin to be able to add the web implementation (suggestion to use x_file originated from @ditman, see [here](https://github.com/flutter/plugins/pull/3253#discussion_r519067323)). Additional interest has been shown to use it in the [image_picker](https://github.com/flutter/plugins/tree/master/packages/image_picker) plugin and [cache_manager](https://github.com/baseflow/flutter_cache_manager) package.

Note: I took the liberty of renaming `XFile` to `CrossFile` since a package with the name "XFile" already exists on pub.dev but with a different purpose.

If this PR is accepted and the package is published, I will make sure to also migrate the [file_selector](https://github.com/flutter/plugins/tree/master/packages/file_selector) plugin to use this new package and delete the internal implementation.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/cross_file/lib/src/web_helpers/web_helpers.dart
Comment (ditman) [inline packages/cross_file/lib/src/web_helpers/web_helpers.dart:1]: This file is missing the copyright notice, can you add it?

File: packages/cross_file/lib/src/types/io.dart
Comment (ditman) [inline packages/cross_file/lib/src/types/io.dart:1]: This file is missing the copyright notice, can you add it?

File: packages/cross_file/lib/src/types/interface.dart
Comment (ditman) [inline packages/cross_file/lib/src/types/interface.dart:1]: This file is missing the copyright notice, can you add it?

File: packages/cross_file/lib/src/types/html.dart
Comment (ditman) [inline packages/cross_file/lib/src/types/html.dart:1]: This file is missing the copyright notice, can you add it?

File: packages/cross_file/lib/src/types/base.dart
Comment (ditman) [inline packages/cross_file/lib/src/types/base.dart:1]: This file is missing the copyright notice, can you add it?

File: packages/cross_file/lib/src/cross_file.dart
Comment (ditman) [inline packages/cross_file/lib/src/cross_file.dart:1]: This file is missing the copyright notice, can you add it?

File: packages/cross_file/lib/cross_file.dart
Comment (ditman) [inline packages/cross_file/lib/cross_file.dart:1]: This file is missing the copyright notice, can you add it?

**Review by ditman — CHANGES_REQUESTED** [review]
Is there a way that we can preserve @jasonpanelli's authorship on this package? Can we maybe add him as a co-author of one of your commits, and then ensure that when we merge he stays as a co-author?

Thanks!

**Comment by mvanbeusekom** [comment]
@googlebot I consent.

**Comment by mvanbeusekom** [comment]
@ditman thanks for the review. I have added the missing copyrights and immediately added @jasonpanelli as co-author. This however resulted in the CLA step failing since Jason now needs to also give his consent (which makes sense, but didn't think about it earlier).

Also I did a `git mv` when copying all files from the file_selector plugin, so if one would run for example `git log --follow packages/cross-file/lib/src/types/base.dart` you can still see the full history including all commits by @jasonpanelli.

**Comment by ditman** [comment]
Thanks for the extra effort @mvanbeusekom!

@jasonpanelli, when you have time, can you please tell googlebot you consent to this change? Thanks!

**Comment by ditman** [comment]
Quick note @mvanbeusekom: there's already some implementations coming for the file_selector plugin [example](https://github.com/flutter/plugins/pull/3265).

I understand that the package name conflicts with something already in pub, and I think CrossFile is a good name for it; but could we maintain the name of the class that this package exports as `XFile`, so we can introduce the new package through the `platform_interface` in a backwards-compatible way?

_(If we decide to go through with naming the class `CrossFile`, we'd need to bump the major version of the platform_interface package interface to introduce the new class name)_

**Comment by mvanbeusekom** [comment]
> Quick note @mvanbeusekom: there's already some implementations coming for the file_selector plugin [example](https://github.com/flutter/plugins/pull/3265).
> 
> I understand that the package name conflicts with something already in pub, and I think CrossFile is a good name for it; but could we maintain the name of the class that this package exports as `XFile`, so we can introduce the new package through the `platform_interface` in a backwards-compatible way?
> 
> _(If we decide to go through with naming the class `CrossFile`, we'd need to bump the major version of the platform_interface package interface to introduce the new class name; this bump cascades to the implementations as well, because those would have an implementation that return an XFile, which then gets migrated to CrossFile)_

This is a good point. I have renamed the classes (and file names) back to `XFile` while keeping the package name `CrossFile`.

**Comment by ditman** [comment]
(I've pinged @jasonpanelli to get his approval!)

**Comment by jasonpanelli** [comment]
@googlebot I consent.

**Comment by ditman** [comment]
Ah yes, I forgot to mention this check:

```
* The name of "lib/x_file.dart", "x_file", should match the name of the package, "cross_file".
  This helps users know what library to import.
```

**Comment by mvanbeusekom** [comment]
> Ah yes, I forgot to mention this check:
> 
> ```
> * The name of "lib/x_file.dart", "x_file", should match the name of the package, "cross_file".
>   This helps users know what library to import.
> ```

Should be fixed now, CI is all green

**Review by ditman — APPROVED** [review]
This is going to be great. If there's any issue with the merge, feel free to add me to the CODEOWNERS for this package!

**Comment by ditman** [comment]
Note to all: `package: cross_file` is now a thing:

https://pub.dev/packages/cross_file

**Pull request merged**
